lib/src/main/java/io/ably/lib/transport/ConnectionManager.java

```diff
@@ -1232,12 +1231,35 @@ private synchronized void onConnected(ProtocolMessage message) {
             requestState(transport, new StateIndication(ConnectionState.failed, e.errorInfo));
             return;
         }
-
         /* indicated connected currentState */
         setSuspendTime();
         requestState(new StateIndication(ConnectionState.connected, error));
     }
 
+    /**
+     * Send all pending messages which are in the queue.
+     * Remove them from the queue once they are sent successfully
+     * Spec: RTN19a
+     * @param resetMessageSerial
+     */
+    private void sendPendingQueueMessages(boolean resetMessageSerial) {
+        //RTN19a
+        if (resetMessageSerial){
+            pendingMessages.resetStartSerial(0);
+        }
+        for (final QueuedMessage queuedMessage : pendingMessages.queue) {
+            try {
+                send(queuedMessage.msg, false, queuedMessage.listener);
```

Comment:
This is incorrect. We don't call the `CompletionListener` just because the `send()` was successful; we call it when we get a response from the server with an `ACK` or `NACK`.
So here, if `send()` succeeds, we will keep the `QueuedMessage` in the queue, and when the server responds we can call the listener. But in the exception case when send() fails we have to decide what to do. Why did it fail? Is it because the connection dropped again? Or is there some other problem?

If we decide that the failure to `send()` means that we're going to give up on sending the message, then here we would remove the `QueuedMessage` from the queue, and call the listener with an error. However, I'm not aware of any circumstances where the `send()` would fail, other than connectivity problems (in which case we'd expect there to be a subsequent resume that triggered another retry attempt).

Reply:
> This is incorrect. We don't call the CompletionListener just because the send() was successful; we call it when we get a response from the server with an ACK or NACK.

Could you please clarify whether this is incorrect because of the following paragraph or because it's not doing what it's supposed to do as described here? As far as I can see, this completion listener will be passed to `PendingMessageQueue` in `sendImpl` if ack required, so ultimately this is going to get its callback from there.  However I realized that message queue inside `PendingMessageQueue` never clears. So I added fa0a4bdec281e5b0245809a4477f2135ad92c533 that will at least clear the queue when fresh messages are added. But the queue remains in memory otherwise, potentially causing memory issues. Shouldn't we remove relevant items from the queue after we received ack or nack from server about them?

> So here, if send() succeeds, we will keep the QueuedMessage in the queue, and when the server responds we can call the listener. But in the exception case when send() fails we have to decide what to do. Why did it fail? Is it because the connection dropped again? Or is there some other problem?

The current behaviour for errors for queued messages for completion listener seem to be inside `PendingQueuedMessages` and connection emits an update with an error if there is an exception with the send. I think we should be removing that relevant message from pending queue if that happens but currently there is no such a mechanism on `PendingMessageQueue` and I cannot see any spec points that describes this case. What would you think should happen in that case?

> If we decide that the failure to send() means that we're going to give up on sending the message, then here we would remove the QueuedMessage from the queue, and call the listener with an error. However, I'm not aware of any circumstances where the send() would fail, other than connectivity problems (in which case we'd expect there to be a subsequent resume that triggered another retry attempt).

As above this should trigger a connection event back to the user, but I'm really not sure what to do in that case. But going through codebase where the message is added into the pending queue, this seems to be happening before transport has sent the message (not even waiting for success of message) I think this will cause another issue where a failed sent message would also be added to the pending queue. Happy to look into this further